### ISSUE TITLE
feat(skore): Add plotly as plotting backend for `Display.plot`

### ISSUE THREAD

**Issue by sylvaincom** [probabl-ai/skore#1079]
### Is your feature request related to a problem? Please describe.

*This was also requested in a user interview. A DS consultant needs plots with interaction to answer questions from clients during a presentation.*

`EstimatorReport` has some plotting features, but only in matplotlib for now. Hence, in the skore UI, the display would not be neat.

### Describe the solution you'd like

We should add plotly as a plotting backend. Kind of what Pandas allows: [Pandas Plotting Backend in Python](https://plotly.com/python/pandas-backend/).

**Comment by sylvaincom** [comment]
Hi @Muhammad-Rebaal, sure thanks, I'll assign this issue to you.

Yes, it would be to allow a plotly equivalent to the matplotlib ones.

**Comment by glemaitre** [comment]
Regarding this issue and associated PR, I would like also that we slow down and think 2 minutes about the internal implementation and the user facing API.

Internally, I think that we should probably think about having a single dispatch mechanism a bit similar to what `skrub` did for pandas/polars: https://github.com/skrub-data/skrub/blob/main/skrub/_dataframe/_common.py

We probably need to think about a single plot to see how many component from matplotlib we are using and if we have a mapping between matplotlib and pyplot. I think that we are not using fancy thing so we should be able to get a middle ground. Maybe the most complex part will be to get the details of the graph (like axis and labels).

In terms of user API, we should make sure that the parameter that we have can be easily dispatch to the matplotlib or plotly implementation and that we don't have something specific to one specific backend. I don't think that we should have a `backend` parameter in every display and instead to have a global config or context manager where we set this option.

**Comment by auguste-probabl** [comment]
Additionally, it appears that plotly supports direct translation from matplotlib: https://github.com/plotly/plotly.py/blob/main/plotly/tools.py#L67
```python
import plotly.tools

plotly.tools.mpl_to_plotly(fig)
```

**Comment by glemaitre** [comment]
Interesting. We should assess until which point this tool help us.

**Comment by glemaitre** [comment]
I quickly check. It is doing part of the job. But you end up needing to tweak the results I think and additional stuff to avoid some warning. In terms of API, it could be weird to make plotly plot and request matplotlib API kwargs though.

**Comment by sylvaincom** [comment]
For the backend, we can have `matplotlib` / `plotly` / `default` arguments where `default` is the one that makes sense considering what is being plotted ; sometimes we can have a “Not implemented” message if the interactivity / static does not make sense for example.

cc @MarieSacksick @glemaitre

**Comment by MarieSacksick** [comment]
To complete, the goal of this issue is to set up this [config manager](https://github.com/probabl-ai/skore/blob/a05b95f/skore/src/skore/_config.py#L81), and only this.  
We will implement little by little the plotly ones. For now, everything we have is matplotlib. All the figures should render "not implemented" if plotly is chosen.

**Comment by divakaivan** [comment]
Hi, I'd like to be assigned to this issue. Thanks

**Comment by divakaivan** [comment]
@MarieSacksick 

### About setting up the config manager ~

 For the [config manager(skore)](https://github.com/probabl-ai/skore/blob/a05b95f/skore/src/skore/_config.py#L81) would extending it in a way similar to how scikit-learn does its own [config manger (scikit-learn)](https://github.com/scikit-learn/scikit-learn/blob/da08f3d99/sklearn/_config.py#L214) be a good way for skore's use case? 

I.e. add like:

```python
_global_config: dict[str, Any] = {
    "show_progress": True,
    "plots": "default", # allowed default, matplotlib, plotly
}
```

and add the `plots` as a param in the `set_config` and `config_context` functions

**Comment by MarieSacksick** [comment]
yes, it's the idea!

**Comment by divakaivan** [comment]
I am trying to wrap my head around the design impl. I looked into `singledispatcher` and how `skrub` uses it and this is what I have so far. I wonder if you agree/disagree with my thought process so far ~

I assume that the goal is to keep the user api simple -> using `.plot()` and if extra options are wanted they are passed into plot.

#### Option 1 - singledispatch on display type (following @glemaitre ’s suggestion inspired by skrub)
 
- the singledispatch works *depending only on the 1st argument passed to it*. In skrub's case the 1st argument is a df/column. In this case it might depend on the type of Display that is shown (i.e. ConfusionMatrixDisplay, RocCurveDisplay)

```python
@singledispatch
def sd_plot(obj, **kwargs):
    raise NotImplementedError("Unsupported type")

@sd_plot.register(ConfusionMatrixDisplay) # repeat for other Displays
def plot_cm(obj, **kwargs):
    if context == "matplotlib": # from context manager
        return plot_cm_matplotlib(obj, **kwargs)
    elif context == "plotly":
        return plot_cm_plotly(obj, **kwargs)
    else:
        raise NotImplementedError(f"Unsupported backend {context}. Available are ['matplotlib', 'plotly']")
```

^^^ Not sure if I like the `if/else` splits.

Usage:

```python
class ConfusionMatrixDisplay:
    def plot(self, **kwargs):
        return sd_plot(self, **kwargs)
```

#### Option 2 - singledispatch on backend figure type

This uses a backend-type-aware dispatcher: we load info about the backend module (`plotly.graph_objs.Figure` or  `matplotlib.Figure`) and dynamically specialize the dispatch

```python
@dispatch
def plot(fig, **kwargs):
    raise NotImplementedError(f"Unsupported backend for fig: {fig}")

@plot.specialize("matplotlib")
def _plot_matplotlib(fig, **kwargs):
    print("Plotting with Matplotlib")

@plot.specialize("plotly")
def _plot_plotly(fig, **kwargs):
    print("Plotting with Plotly")
```

I spend some time translating the skrub dispatcher to skore's use case but I think the version I created turned out to be not such a good idea (?) because we wouldn't want to create the graph first and then pass it to the dispatcher.


Edit: the design explanation is pretty good in the skrub code [here](https://github.com/skrub-data/skrub/blob/main/skrub/_dispatch.py)

<details>
  <summary>Click to expand</summary>

```python
from dataclasses import dataclass
from types import MappingProxyType, ModuleType
from typing import Any, Dict, Tuple
from functools import singledispatch

@dataclass
class PlotModuleInfo:
    name: str
    module: ModuleType
    types: Dict[str, Tuple[Any]]

def _load_plot_module_info(name: str) -> PlotModuleInfo:
    if name == "matplotlib":
        import matplotlib.pyplot as plt
        return PlotModuleInfo(
            **{
                "name": "matplotlib",
                "module": plt,
                "types": MappingProxyType(
                    {
                        "Figure": (plt.Figure,),
                    }
                )
            }
        )
    if name == "plotly":
        import plotly.graph_objs as go
        import plotly.express as px
        return PlotModuleInfo(
            **{
                "name": "plotly",
                "module": px,
                "types": MappingProxyType(
                    {
                        "Figure": (go.Figure,),
                    }
                )
            }
        )
    raise KeyError(
        f"Unknown plotting backend module: {name}."
        "Available modules are ['matplotlib', 'plotly']."
    )

def dispatch(function):
    dispatched = singledispatch(function)

    def specialize(module_name, *, argument_type=None):

        module_info = _load_plot_module_info(module_name)
        if argument_type is None:
            argument_type = list(module_info.types.keys())
        elif isinstance(argument_type, str):
            argument_type = (argument_type,)

        def decorator(specialized_impl):
            types_to_register = set()
            for type_name in argument_type:
                types_to_register.update(module_info.types[type_name])
            for module_type in types_to_register:
                dispatched.register(module_type, specialized_impl)
            return specialized_impl

        return decorator

    dispatched.specialize = specialize
    return dispatched

############ USING IT ########################

@dispatch
def plot(fig, **kwargs):
    raise NotImplementedError(
        f"Plotting not implemented for figure: {fig}. Available are ['matplotlib', 'plotly']."
    )

@plot.specialize("matplotlib")
def _plot_matplotlib(classs, **kwargs):
    print("Plotting with Matplotlib")

@plot.specialize("plotly")
def _plot_plotly(classs, **kwargs):
    print("Plotting with Plotly")


import matplotlib.pyplot as plt
import plotly.graph_objs as go

plt_fig = plt.figure()
plotly_fig = go.Figure()
plot(plt_fig) # -> Plotting with Matplotlib
plot(plotly_fig) # -> Plotting with Plotly
```

</details>

**Comment by MarieSacksick** [comment]
I think I like the second option best, because I find it more lightweight and readible.

**Comment by auguste-probabl** [comment]
Let's keep this as a tracking issue for re-implementing every plot in Plotly, or open a new one for that purpose.

**Comment by glemaitre** [comment]
So we can open sub-isues in this issue to track the progress.

**Comment by auguste-probabl** [comment]
It will be tricky to replicate the seaborn setup we have now.
